pkg/discovery/util/node_util.go

```diff
@@ -217,3 +218,21 @@ func DetectHARole(node *api.Node) bool {
 	}
 	return isHANode
 }
+
+func MapNodePoolToNodeNames(nodes []*v1.Node) map[string]sets.String {
+	glog.V(4).Info("mapping node pools to nodes.")
+	nodePools := make(map[string]sets.String)
+	for _, node := range nodes {
+		allPools := DetectNodePools(node)
+		for _, pool := range allPools.UnsortedList() {
+			if nodePools[pool] == nil {
+				nodePools[pool] = sets.NewString(node.Name)
+			} else {
+				nodePools[pool].Insert(node.Name)
+			}
+		}
+	}
+	glog.V(4).Infof("Found %+v node pool keys.", len(nodePools))
+	glog.V(5).Info(nodePools)
+	return nodePools
+}
```

Comment:
How about when the `nodePools` map is empty, you return a map with the key: **Cluster**, value: a set of all nodes in the cluster (i.e., a fake pool with all nodes in the cluster). Then you don't need to have the logic later on to check if there are node pools or not.

Reply:
I like this potentially being reusable and if I add a Cluster key, that would be an implementation detail the caller would have to know about prior to using this function.

Reply:
adding the cluster key to the calling code.